refextract/references/pdf.py

```diff
@@ -48,40 +47,119 @@ def extract_texkeys_from_pdf(pdf_file):
             destinations = pdf.getNamedDestinations()
+            urls = extract_urls(pdf)
         except Exception:
             LOGGER.debug(u"PDF: Internal PyPDF2 error, no TeXkeys returned.")
             return []
         # not all named destinations point to references
-        refs = [dest for dest in destinations.items()
-                if re_reference_in_dest.match(dest[0])]
+        refs = [
+            dest for dest in destinations.items() if re_reference_in_dest.match(dest[0])
+        ]
         try:
             if _destinations_in_two_columns(pdf, refs):
                 LOGGER.debug(u"PDF: Using two-column layout")
 
                 def sortfunc(dest_couple):
-                    return _destination_position(pdf, dest_couple[1])
+                    return dest_couple[1]
+
+                def uri_sortfunc(uri_obj):
+                    return uri_obj[1]
 
             else:
                 LOGGER.debug(u"PDF: Using single-column layout")
 
                 def sortfunc(dest_couple):
-                    (page, _, ypos, xpos) = _destination_position(
-                        pdf, dest_couple[1])
+                    page, _, ypos, xpos = dest_couple[1]
                     return (page, ypos, xpos)
 
-            refs.sort(key=sortfunc)
-            # extract the TeXkey from the named destination name
-            return [re_reference_in_dest.match(destname).group(1)
-                    for (destname, _) in refs]
+                def uri_sortfunc(uri_obj):
+                    page, _, ypos, xpos = uri_obj[1]
+                    return (page, ypos, xpos)
+
+            refs = [(dest[0], _destination_position(pdf, dest[1])) for dest in refs]
+            refs.sort(key=uri_sortfunc)
+            urls = [
+                (uri["/A"]["/URI"], _uri_destination_position(pdf, uri)) for uri in urls
+            ]
+            urls.sort(key=uri_sortfunc)
+            texkey_url_list = []
+            for nb, ref in enumerate(refs):
+                current_texkey_urls_dict = {}
+                current_texkey_urls_dict["texkey"] = re_reference_in_dest.match(
+                    ref[0]
+                ).group(1)
+                if nb < len(refs) - 1:
+                    next_reference_data = refs[nb + 1]
+                    matched_urls_for_reference, urls = _match_url_with_references(
+                        urls, ref, next_reference_data
+                    )
+                else:
+                    matched_urls_for_reference = _match_last_reference(urls, ref)
+                if matched_urls_for_reference:
+                    current_texkey_urls_dict["urls"] = matched_urls_for_reference
+                texkey_url_list.append(current_texkey_urls_dict)
+
+            return texkey_url_list
         except Exception:
             LOGGER.debug(u"PDF: Impossible to determine layout, no TeXkeys returned")
             return []
 
 
+def _match_url_with_references(urls_to_match, reference, next_reference):
```

Comment:
```suggestion
def _match_urls_with_reference(urls_to_match, reference, next_reference):
```
as you associate possibly multiple urls to a single ref here?